File: crates/neuronprompter-web/frontend/src/api/dialogs.ts
/**
 * Native file dialog bridge.
 *
 * In desktop mode (wry), calls the web endpoint which opens a native OS dialog
 * via the rfd crate. In browser mode, returns null (caller should use
 * browser-native download/upload instead).
 */

const BASE = "/api/v1/web/dialog";

interface DialogFilter {
  name: string;
  extensions: string[];
}

interface DialogRequest {
  title?: string;
  filters?: DialogFilter[];
  default_name?: string;
}

interface DialogResult {
  path: string | null;
}

async function postDialog(endpoint: string, body: DialogRequest = {}): Promise<string | null> {
  try {
    const res = await fetch(`${BASE}/${endpoint}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (!res.ok) return null;
    const result: DialogResult = await res.json();
    return result.path;
  } catch {
    return null;
  }
}

/** Opens a native save-file dialog. Returns the chosen path or null. */
export async function showSaveDialog(
  title: string,
  filters: DialogFilter[],
  defaultName?: string,
): Promise<string | null> {
  return postDialog("save", { title, filters, default_name: defaultName });
}

/** Opens a native open-file dialog. Returns the chosen path or null. */
export async function showOpenFileDialog(
  title: string,
  filters: DialogFilter[],
): Promise<string | null> {
  return postDialog("open-file", { title, filters });
}

/** Opens a native directory picker. Returns the chosen path or null. */
export async function showOpenDirDialog(): Promise<string | null> {
  return postDialog("open-dir");
}

File: crates/neuronprompter-web/src/handlers/dialogs.rs
// =============================================================================
// Native file dialog endpoints.
//
// Opens OS-native save/open file dialogs via the rfd crate and returns the
// selected path as JSON. Falls back gracefully when compiled without the
// gui feature (rfd unavailable) by returning null.
// =============================================================================

use std::sync::Arc;

use axum::Json;
use axum::extract::State;
use serde::{Deserialize, Serialize};

use crate::WebState;

#[derive(Deserialize)]
pub struct DialogRequest {
    /// Dialog window title.
    pub title: Option<String>,
    /// File extension filters, e.g. `["json"]` or `["db", "sqlite"]`.
    pub filters: Option<Vec<DialogFilter>>,
    /// Default file name for save dialogs.
    pub default_name: Option<String>,
}

#[derive(Deserialize)]
pub struct DialogFilter {
    pub name: String,
    pub extensions: Vec<String>,
}

#[derive(Serialize)]
pub struct DialogResult {
    /// The selected file/directory path, or null if the user cancelled.
    pub path: Option<String>,
}

/// POST /api/v1/web/dialog/save
///
/// Opens a native save-file dialog and returns the chosen path.
pub async fn save_dialog(
    State(_state): State<Arc<WebState>>,
    Json(body): Json<DialogRequest>,
) -> Json<DialogResult> {
    Json(DialogResult {
        path: open_save_dialog(&body),
    })
}

/// POST /api/v1/web/dialog/open-file
///
/// Opens a native open-file dialog and returns the chosen path.
pub async fn open_file_dialog(
    State(_state): State<Arc<WebState>>,
    Json(body): Json<DialogRequest>,
) -> Json<DialogResult> {
    Json(DialogResult {
        path: open_file_dialog_impl(&body),
    })
}

/// POST /api/v1/web/dialog/open-dir
///
/// Opens a native directory picker and returns the chosen path.
pub async fn open_dir_dialog(State(_state): State<Arc<WebState>>) -> Json<DialogResult> {
    Json(DialogResult {
        path: open_dir_dialog_impl(),
    })
}

// ---------------------------------------------------------------------------
// rfd implementations (compiled only with gui feature)
// ---------------------------------------------------------------------------

#[cfg(feature = "gui")]
fn open_save_dialog(req: &DialogRequest) -> Option<String> {
    let mut dialog = rfd::FileDialog::new();
    if let Some(title) = &req.title {
        dialog = dialog.set_title(title);
    }
    if let Some(name) = &req.default_name {
        dialog = dialog.set_file_name(name);
    }
    if let Some(filters) = &req.filters {
        for f in filters {
            let exts: Vec<&str> = f.extensions.iter().map(String::as_str).collect();
            dialog = dialog.add_filter(&f.name, &exts);
        }
    }
    dialog.save_file().map(|p| p.display().to_string())
}

#[cfg(feature = "gui")]
fn open_file_dialog_impl(req: &DialogRequest) -> Option<String> {
    let mut dialog = rfd::FileDialog::new();
    if let Some(title) = &req.title {
        dialog = dialog.set_title(title);
    }
    if let Some(filters) = &req.filters {
        for f in filters {
            let exts: Vec<&str> = f.extensions.iter().map(String::as_str).collect();
            dialog = dialog.add_filter(&f.name, &exts);
        }
    }
    dialog.pick_file().map(|p| p.display().to_string())
}

#[cfg(feature = "gui")]
fn open_dir_dialog_impl() -> Option<String> {
    rfd::FileDialog::new()
        .set_title("Select Directory")
        .pick_folder()
        .map(|p| p.display().to_string())
}

// ---------------------------------------------------------------------------
// Fallbacks when gui feature is not compiled
// ---------------------------------------------------------------------------

#[cfg(not(feature = "gui"))]
fn open_save_dialog(_req: &DialogRequest) -> Option<String> {
    None
}

#[cfg(not(feature = "gui"))]
fn open_file_dialog_impl(_req: &DialogRequest) -> Option<String> {
    None
}

#[cfg(not(feature = "gui"))]
fn open_dir_dialog_impl() -> Option<String> {
    None
}
